### PR TITLE
Query auth and bonds code freeze

### PR DESCRIPTION
**Pull request by wahlbergkyle**
# Description

Adding Query Authentication and updated Bonds.

## Needs to change
I believe there's a better way to use Red's shadeadmin repo than the package I made, but I failed to figure it out while making this.
Additionally, old integration tests that use the oracle will fail as it was only a partial change to get Bonds functional with the new oracle standard.

### CODE REVIEW

File: contracts/bonds/Cargo.toml [hunk 1]
```diff
@@ -41,8 +41,18 @@ serde = { version = "1.0.103", default-features = false, features = ["derive"] }
...
+shade_admin = { git = "https://github.com/securesecrets/shadeadmin.git", tag = "v1.0"}
+shade-oracles = { git = "https://github.com/securesecrets/shade-oracle.git", tag = "0.11"}
+fadroma = { git = "https://github.com/hackbg/fadroma", branch = "v100", features = [] }
 query-authentication = {git = "https://github.com/securesecrets/query-authentication", tag = "v1.3.0"}
 
 [dev-dependencies]
 mockall = "0.10.2"
-mockall_double = "0.2.0"
\ No newline at end of file
+mockall_double = "0.2.0"
```
Comment (DrPresident) [inline contracts/bonds/Cargo.toml:52]: I think neither of these are used, I think they were added in back when we didn't have a good "mock chain" solution and they have just kind of lingered. Try building without them and remove them if it works.

File: contracts/bonds/src/contract.rs [hunk 2]
```diff
@@ -33,7 +33,7 @@ pub fn init<S: Storage, A: Api, Q: Querier>(
 ) -> StdResult<InitResponse> {
     let state = Config {
         limit_admin: msg.limit_admin,
-        admin: msg.admin,
+        shade_admins: msg.shade_admins,
```
Comment (DrPresident) [inline contracts/bonds/src/contract.rs:36]: Can we make this singular `shade_admin`? just makes more sense as its a contract & not a list

File: contracts/bonds/src/handle.rs [hunk 3]
```diff
@@ -12,10 +12,12 @@ use secret_toolkit::{
     utils::{HandleCallback, Query},
 };
 
+use shade_admin::admin::{ValidateAdminPermissionResponse, QueryMsg};
+
+use shade_oracles::{router::{QueryMsg::GetPrice}, common::OraclePrice};
```
Comment (DrPresident) [inline contracts/bonds/src/handle.rs:17]:
```suggestion
use shade_oracles::{router::QueryMsg::GetPrice, common::OraclePrice};
```

File: contracts/bonds/src/handle.rs [hunk 4]
```diff
@@ -109,9 +115,19 @@ pub fn try_update_config<S: Storage, A: Api, Q: Querier>(
...
+        contract_address: cur_config.contract.to_string(), 
+        admin_address: env.message.sender.to_string(), 
+    }.query(
+        &deps.querier,
+        cur_config.shade_admins.code_hash,
+        cur_config.shade_admins.address,
+    )?;
+
+    if admin_response.error_msg.is_some() {
+        return Err(not_admin())
```
Comment (DrPresident) [inline contracts/bonds/src/handle.rs:130]: We should have this all in a single function like `shade_admin::is_admin(contract, admin, shade_admin)`

File: contracts/bonds/src/handle.rs [hunk 5]
```diff
@@ -758,7 +745,7 @@ pub fn amount_to_issue<S: Storage, A: Api, Q: Querier>(
     err_issued_price: Uint128,
 ) -> StdResult<(Uint128, Uint128, Uint128, Uint128)> {
     let mut disc = discount;
-    let mut collateral_price = oracle(&deps, collateral_asset.token_info.symbol.clone())?; // Placeholder for Oracle lookup
+    let mut collateral_price = oracle(&deps, collateral_asset.token_info.symbol.clone())?;
```
Comment (DrPresident) [inline contracts/bonds/src/handle.rs:748]: I think "collateral" is the wrong terminology here as its not a loan & it isn't leveraged, it's more of a purchase

File: contracts/bonds/src/query.rs [hunk 6]
```diff
@@ -37,10 +37,10 @@ pub fn account<S: Storage, A: Api, Q: Querier>(
     )?;
     match authorized {
         query_auth::QueryAnswer::ValidatePermit { user, is_revoked } => {
-            if is_revoked!=false {
+            if is_revoked!=true {
```
Comment (DrPresident) [inline contracts/bonds/src/query.rs:40]:
```suggestion
            if is_revoked != true {
```

File: contracts/bonds/Cargo.toml [hunk 7]
```diff
@@ -41,8 +41,18 @@ serde = { version = "1.0.103", default-features = false, features = ["derive"] }
...
+fadroma = { branch = "v100", git = "https://github.com/hackbg/fadroma.git", features= ["ensemble"]  }
```
Comment (FloppyDisck) [inline contracts/bonds/Cargo.toml:51]:
```suggestion
fadroma = { branch = "v100", commit = 76867e0, git = "https://github.com/hackbg/fadroma.git", features= ["ensemble"]  }
```

File: contracts/bonds/src/tests/query.rs [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    bonds, 
+    snip20::{helpers::Snip20Asset}, 
+    query_auth::{self, PermitData, QueryPermit},
+};
+
+use query_authentication::transaction::{PermitSignature, PubKey};
+
+use cosmwasm_math_compat::Uint128;
+
+pub fn query_no_opps(
```
Comment (DrPresident) [inline contracts/bonds/src/tests/query.rs:14]: Does this need the `#[test]` flag?

File: contracts/bonds/src/query.rs (hunk 6 above)
Comment (FloppyDisck) [inline contracts/bonds/src/query.rs]:
```suggestion
            if !is_revoked {
```

**Review by DrPresident — APPROVED** [review]
Looks like warnings are cleaned up, some small adjustment suggestions as well as it might be worth running a format on this to touch up some of the spacing around operators and such.

We should also have functions for the shade_admin queries to clean that up as its a multi-step authentication thats being done in multiple places.

File: contracts/bonds/Cargo.toml (hunk 1 above)
Comment (wahlbergkyle) [reply contracts/bonds/Cargo.toml:52]: It works without them, updated and removed

File: contracts/bonds/src/tests/query.rs (hunk 8 above)
Comment (wahlbergkyle) [reply contracts/bonds/src/tests/query.rs:14]: No, it's a function that's called by another test as opposed to a test on its own

File: contracts/bonds/src/handle.rs (hunk 4 above)
Comment (wahlbergkyle) [reply contracts/bonds/src/handle.rs:130]: Hopefully in the next release tag we can, I agree it'd be easier

**Review by FloppyDisck — CHANGES_REQUESTED** [review]
Missing some testing but its fine, please review the comments I left

File: contracts/bonds/Cargo.toml [hunk 9]
```diff
@@ -41,8 +41,16 @@ serde = { version = "1.0.103", default-features = false, features = ["derive"] }
 snafu = { version = "0.6.3" }
 chrono = "0.4.19"
 time = "0.1.44"
+admin = { git = "https://github.com/securesecrets/shadeadmin.git", tag = "v1.0" }
+shade_admin = { git = "https://github.com/securesecrets/shadeadmin.git", tag = "v1.0"}
+shade-oracles = { git = "https://github.com/securesecrets/shade-oracle.git", tag = "0.11"}
+fadroma = { git = "https://github.com/hackbg/fadroma", branch = "v100", features = [] }
```
[the remark this answers is not in the document]
Comment (wahlbergkyle) [reply contracts/bonds/Cargo.toml]: Yeah it's gone now

File: contracts/bonds/src/state.rs [hunk 10]
```diff
@@ -4,13 +4,9 @@ use cosmwasm_storage::{
...
-    contract_interfaces::{
-        bonds::{
-            Account, BondOpportunity, Config,
-        },
-        snip20::helpers::Snip20Asset,
-    },
+use shade_protocol::contract_interfaces::{
+    bonds::{Account, BondOpportunity, Config},
+    snip20::helpers::Snip20Asset,
 };
```
Comment (SissonJ) [inline contracts/bonds/src/state.rs:11]: There's a way to remove the necessity of a state file, @FloppyDisck do we want this change going forward? or is it more of a "nice to have"?

File: contracts/bonds/Cargo.toml [hunk 11]
```diff
@@ -41,8 +41,18 @@ serde = { version = "1.0.103", default-features = false, features = ["derive"] }
...
```
[the remark this answers is not in the document]
Comment (SissonJ) [reply contracts/bonds/Cargo.toml]: I stand outdated

File: contracts/bonds/src/tests/query.rs [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+    minting_bond: Option<bool>
+) -> () {
+    let query: bonds::QueryAnswer = chain.query(
+        bonds.address.clone(),
+        &bonds::QueryMsg::BondOpportunities {  }
+    ).unwrap();
+
+    match query {
+        bonds::QueryAnswer::BondOpportunities { bond_opportunities, .. } => {
+            if issuance_limit.is_some() {
```
[the remark this answers is not in the document]
Comment (wahlbergkyle) [reply contracts/bonds/src/tests/query.rs:52]: Yessir

**Review by FloppyDisck — APPROVED** [review]

File: contracts/bonds/src/handle.rs (hunk 5 above)
Comment (wahlbergkyle) [reply contracts/bonds/src/handle.rs:748]: The Olympus Dao github suggests they refer to the assets as "Deposit" and "Redeem", respectively, whereas we've been using "Collateral" and "Issued"

Comment (wahlbergkyle) [reply contracts/bonds/src/handle.rs]: Updated to "Deposit" everywhere, no longer collateral

**Pull request merged**